proto/p4/p4runtime.proto

```diff
@@ -356,6 +357,29 @@ message CloneSessionEntry {
   uint64 packet_length_bytes = 4;
 }
 
+//------------------------------------------------------------------------------
+// For writing and reading matches in a parser value set. A state transition
+// on an empty value set will never be taken. The number of matches must be at
+// most the size of the value set as specified by the @size annotation in P4
+// program.
+//
+// For Write Requests:
+//    INSERT will write the given matches in the repeated field to the value
+//    set.
+//    MODIFY will replace all matches in the value set with the given matches
+//    in the repeated field.
+//    DELETE should only specify the value_set_id and will delete all matches
+//    in the value set.
+//
+// For Read Requests:
+//    All matches for all value-set entries if value_set_id = 0
+//    All matches of the value-set if a valid value_set_id is specified
+//    The 'match' field must never be set in the ReadRequest
+message ValueSetEntry {
+  uint32 value_set_id = 1;
+  repeated FieldMatch match = 2;
```

Comment:
FieldMatch should be MatchField?

Bigger question: A table can have one or more fields that it matches on.

A value_set could potentially be allowed to match on multiple fields, too, e.g. if the transition select statement gave a list of fields, like the one in this example P4_16 program: https://github.com/p4lang/p4c/blob/master/testdata/p4_16_samples/parser-locals2.p4#L76-L80

Does this proposed API enable treating that as a set of "_record_ of field match criteria" (I don't know the right name to use instead of _record_ there -- I simply mean that there could be multiple field match criteria in a single element of the "value set")?

Reply:
That's a separate question that applies to several other places in P4Runtime and is related to the discussion on type-serialization for control-plane consumption. A value_set should probably not be allowed to match on multiple expressions like a table can - unless there is a use-case. However a single match expression can be a more complex type than just a bitstring, e.g. a tuple of header fields (this applies to both tables & PVS). This question should definitely be answered in some form before the P4Runtime v1.0 release. However we should answer this question generally for the entire P4Runtime API and there is no reason not to move forward with this PR in the mean time.

Reply:
Makes sense.  Thanks.

Reply:
Actually, rereading your comment confuses me again.  You say "A value_set should probably not be allowed to match on multiple expressions like a table can" and later "a single match expression can be a more complex type than just a bitstring, e.g. a tuple of header fields (this applies to both tables & PVS)".

The latter quote makes sense to me, and the parser value set example I linked to in my earlier comment is exactly that -- a select statement with a tuple of header fields.

How can a table be more general than that, by a "match on multiple expressions"?  Are you referring to the ability of table keys in P4_16 to be expressions like "meta.field_name & 0xff00"?  If not, then I don't understand the distinction you are making.

Reply:
This is how I see things:
- a table is matching on several separate expressions, each one is a different entry in the "key" property and can have its own static match kind / type. For example:
```
table t {
    key = { expr1 : exact; expr2 : ternary; }
    // ...
}
```
Table `t` matches on 2 expressions and each entry programmed by P4Runtime takes 2 `FieldMatch` values, one for each expression. Most of the time, an expression is just a bitstring (.e.g header field), but we could also match on tuples in a table:
```
table t {
    key = { hdr.ethernet.srcAddr : exact; { hdr.ipv4.srcAddr, hdr.ipv4.dstAddr} : ternary; }
    // ...
}
```
- a parser value set matches on a single expression (which can be of type bitstring, tuple, struct, etc.), with a match kind / type which is dynamic (and not static). Single expression translates into a single `FieldMatch` value in P4Runtime (whereas a table has one per entry in its key).

So in a nutshell I was referring to a table's ability to have several match components in its key (each with its own static match kind), while a PVS has a single match component (with a dynamic match kind).

We will need some work to support complex expressions (such as tuple) in P4Runtime in tables, parser value sets, digest, etc... It is fine IMO to postpone this work for now and merge that proposal as it is. We will need to support these more complex types uniformly in the API. Probably in P4Runtime v1.0.

Reply:
FYI, my comments and questions here are _not_ because I am trying to delay this change getting merged as is.  More just trying to understand the plans of what will be possible later.  Thanks.

Reply:
@jafingerhut my bad, that's not what I meant to imply.
I think serialization of arbitrary P4 types is probably the biggest outstanding question that has to be answered before the v1.0. Maybe that should be the main focus of our next WG meeting?